File: crates/buzz-core/src/task.rs
//! Task lifecycle enums shared across crates.
//!
//! These live in `buzz-core` (zero I/O deps) so the DB layer, the relay HTTP
//! surface, and future clients agree on one spelling of a task's status and of
//! the lifecycle events that status changes record.
//!
//! Tasks are durable work items owned by a human or a harness agent. They are
//! deliberately unrelated to `buzz-workflow`, which models the scheduled
//! execution engine.

use std::fmt;
use std::str::FromStr;

/// Where a task sits in its lifecycle.
///
/// The spelling of each variant is the value stored in `tasks.status` and
/// pinned by that column's `CHECK` constraint, so adding a variant here
/// requires a migration.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TaskStatus {
    /// Accepted but not started.
    Todo,
    /// Actively being worked.
    InProgress,
    /// Cannot proceed until something else resolves.
    Blocked,
    /// Finished successfully.
    Done,
    /// Abandoned without completion.
    Cancelled,
}

impl TaskStatus {
    /// Canonical string representation (matches the `tasks.status` CHECK).
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Todo => "todo",
            Self::InProgress => "in_progress",
            Self::Blocked => "blocked",
            Self::Done => "done",
            Self::Cancelled => "cancelled",
        }
    }

    /// Whether the task has left the working set (done or cancelled).
    pub fn is_closed(&self) -> bool {
        matches!(self, Self::Done | Self::Cancelled)
    }

    /// Whether `tasks.done_at` must carry a timestamp in this status.
    ///
    /// `done_at` is the completion timestamp, so it is set for `Done` and only
    /// for `Done` — cancelling a task closes it without completing it. The
    /// database enforces the same equivalence via
    /// `chk_tasks_done_at_matches_status`; this keeps the write path from
    /// having to learn that constraint by failing it.
    pub fn requires_done_at(&self) -> bool {
        matches!(self, Self::Done)
    }
}

impl fmt::Display for TaskStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for TaskStatus {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "todo" => Ok(Self::Todo),
            "in_progress" => Ok(Self::InProgress),
            "blocked" => Ok(Self::Blocked),
            "done" => Ok(Self::Done),
            "cancelled" => Ok(Self::Cancelled),
            other => Err(format!("unknown task status: {other:?}")),
        }
    }
}

/// A row in the append-only `task_events` log.
///
/// Stored as free `TEXT` rather than a database enum so a new action can ship
/// across a rolling upgrade without a migration; this enum is the set the
/// relay itself writes.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TaskAction {
    /// The task was created.
    Created,
    /// `status` moved from one value to another.
    StatusChanged,
    /// `assignee_pubkey` changed.
    Assigned,
    /// A human or agent left a comment.
    Commented,
    /// `title` changed.
    TitleChanged,
    /// An agent persisted its summary of the task. At most one per task.
    SummaryPersisted,
}

impl TaskAction {
    /// Canonical string representation (matches `task_events.action`).
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Created => "created",
            Self::StatusChanged => "status_changed",
            Self::Assigned => "assigned",
            Self::Commented => "commented",
            Self::TitleChanged => "title_changed",
            Self::SummaryPersisted => "summary_persisted",
        }
    }

    /// Whether at most one event with this action may exist per task.
    ///
    /// Mirrors the partial unique index `idx_task_events_one_summary_per_task`.
    pub fn is_singleton_per_task(&self) -> bool {
        matches!(self, Self::SummaryPersisted)
    }
}

impl fmt::Display for TaskAction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for TaskAction {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "created" => Ok(Self::Created),
            "status_changed" => Ok(Self::StatusChanged),
            "assigned" => Ok(Self::Assigned),
            "commented" => Ok(Self::Commented),
            "title_changed" => Ok(Self::TitleChanged),
            "summary_persisted" => Ok(Self::SummaryPersisted),
            other => Err(format!("unknown task action: {other:?}")),
        }
    }
}

/// The lifecycle event a status change records, or `None` when the requested
/// status is the one the task already has.
///
/// A `PATCH` that restates the current status is idempotent: it must not append
/// a `status_changed` row claiming a transition that did not happen, otherwise
/// a client retry inflates the task's history.
pub fn status_change_action(from: TaskStatus, to: TaskStatus) -> Option<TaskAction> {
    (from != to).then_some(TaskAction::StatusChanged)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn status_round_trips_through_its_canonical_spelling() {
        for status in [
            TaskStatus::Todo,
            TaskStatus::InProgress,
            TaskStatus::Blocked,
            TaskStatus::Done,
            TaskStatus::Cancelled,
        ] {
            assert_eq!(
                status.as_str().parse::<TaskStatus>(),
                Ok(status),
                "{status} must survive a string round trip"
            );
        }
        assert!("in-progress".parse::<TaskStatus>().is_err());
        assert!("DONE".parse::<TaskStatus>().is_err());
    }

    #[test]
    fn action_round_trips_through_its_canonical_spelling() {
        for action in [
            TaskAction::Created,
            TaskAction::StatusChanged,
            TaskAction::Assigned,
            TaskAction::Commented,
            TaskAction::TitleChanged,
            TaskAction::SummaryPersisted,
        ] {
            assert_eq!(action.as_str().parse::<TaskAction>(), Ok(action));
        }
        assert!("summary".parse::<TaskAction>().is_err());
    }

    #[test]
    fn a_real_status_change_records_status_changed() {
        assert_eq!(
            status_change_action(TaskStatus::Todo, TaskStatus::InProgress),
            Some(TaskAction::StatusChanged)
        );
        assert_eq!(
            status_change_action(TaskStatus::Blocked, TaskStatus::Done),
            Some(TaskAction::StatusChanged)
        );
        // Reopening is a transition like any other — the log is append-only,
        // so it records the move rather than rewriting the earlier one.
        assert_eq!(
            status_change_action(TaskStatus::Done, TaskStatus::Todo),
            Some(TaskAction::StatusChanged)
        );
    }

    #[test]
    fn restating_the_current_status_records_nothing() {
        for status in [
            TaskStatus::Todo,
            TaskStatus::InProgress,
            TaskStatus::Blocked,
            TaskStatus::Done,
            TaskStatus::Cancelled,
        ] {
            assert_eq!(
                status_change_action(status, status),
                None,
                "restating {status} must not append a status_changed row"
            );
        }
    }

    #[test]
    fn done_at_is_required_exactly_for_done() {
        // Pins the Rust side of `chk_tasks_done_at_matches_status`: cancelled
        // closes a task without completing it, so it carries no done_at.
        assert!(TaskStatus::Done.requires_done_at());
        for status in [
            TaskStatus::Todo,
            TaskStatus::InProgress,
            TaskStatus::Blocked,
            TaskStatus::Cancelled,
        ] {
            assert!(
                !status.requires_done_at(),
                "{status} must not carry a completion timestamp"
            );
        }
    }

    #[test]
    fn closed_covers_both_terminal_statuses() {
        assert!(TaskStatus::Done.is_closed());
        assert!(TaskStatus::Cancelled.is_closed());
        assert!(!TaskStatus::Todo.is_closed());
        assert!(!TaskStatus::InProgress.is_closed());
        assert!(!TaskStatus::Blocked.is_closed());
    }

    #[test]
    fn only_the_summary_action_is_capped_at_one_per_task() {
        assert!(TaskAction::SummaryPersisted.is_singleton_per_task());
        for action in [
            TaskAction::Created,
            TaskAction::StatusChanged,
            TaskAction::Assigned,
            TaskAction::Commented,
            TaskAction::TitleChanged,
        ] {
            assert!(!action.is_singleton_per_task());
        }
    }
}

File: mobile/lib/shared/tasks/task.dart
/// Task and task-event models plus the request-payload builders for the
/// relay's `/api/tasks` surface.
///
/// The wire format is deliberately asymmetric, so the encode and decode paths
/// below are not mirror images:
///
/// * **Outbound** `due_at` is an RFC 3339 string — `api::tasks` deserializes it
///   into a `chrono::DateTime<Utc>`.
/// * **Inbound** `due_at`, `created_at`, `updated_at`, `done_at` and
///   `archived_at` are Unix **seconds**, because `task_json` emits
///   `value.timestamp()`.
///
/// Keep this file free of Flutter and network imports: everything here is a
/// pure function or a value type so it can be unit-tested without a harness.
library;

import 'package:flutter/foundation.dart';

/// Longest title the relay accepts, counted the way it counts.
///
/// `api::tasks::validate_title` uses `chars().count()` and the table's
/// `CHECK (length(title) BETWEEN 1 AND 200)` counts characters too, so the
/// client-side guard must count Unicode scalar values — `String.runes`, not
/// `String.length` (UTF-16 code units) and not grapheme clusters.
const maxTaskTitleChars = 200;

/// Number of characters the relay will count in [title].
int taskTitleLength(String title) => title.runes.length;

/// Lifecycle state of a task, mirroring `buzz_core::task::TaskStatus`.
enum TaskStatus {
  /// Open and unstarted.
  todo('todo'),

  /// Being worked on.
  inProgress('in_progress'),

  /// Waiting on something else.
  blocked('blocked'),

  /// Finished.
  done('done'),

  /// Abandoned.
  cancelled('cancelled');

  const TaskStatus(this.wireValue);

  /// The string the relay reads and writes for this status.
  final String wireValue;

  /// Parses a relay status string.
  ///
  /// An unrecognised value degrades to [TaskStatus.todo] rather than throwing:
  /// a client built before a status was added must still render the rest of a
  /// task list during a rolling upgrade.
  static TaskStatus fromWire(String? raw) {
    for (final status in TaskStatus.values) {
      if (status.wireValue == raw) return status;
    }
    return TaskStatus.todo;
  }
}

/// The two task-event actions a client may post directly.
///
/// `api::tasks::append_task_event` rejects every other action, because
/// lifecycle entries (`created`, `status_changed`, `assigned`,
/// `title_changed`) are derived from the mutation that caused them — accepting
/// one from a caller would let it fabricate a history that never happened.
enum TaskEventAction {
  /// A free-text comment.
  commented('commented'),

  /// The task's single persisted summary. The relay's
  /// `idx_task_events_one_summary_per_task` partial unique index rejects a
  /// second one with a 400.
  summaryPersisted('summary_persisted');

  const TaskEventAction(this.wireValue);

  /// The string the relay reads for this action.
  final String wireValue;
}

/// Decodes a relay timestamp field (Unix seconds) into local time.
DateTime? _dateFromSeconds(Object? value) {
  if (value is! int) return null;
  return DateTime.fromMillisecondsSinceEpoch(value * 1000, isUtc: true);
}

String? _stringOrNull(Object? value) => value is String ? value : null;

/// A task as returned by `GET`/`POST`/`PATCH /api/tasks`.
@immutable
class Task {
  /// Creates a task record.
  const Task({
    required this.id,
    required this.title,
    required this.status,
    required this.priority,
    required this.createdAt,
    required this.updatedAt,
    this.channelId,
    this.createdBy,
    this.assignee,
    this.parentTaskId,
    this.body,
    this.source,
    this.sourceRef,
    this.dueAt,
    this.doneAt,
    this.archivedAt,
  });

  /// Parses one relay task object.
  factory Task.fromJson(Map<String, dynamic> json) {
    final id = _stringOrNull(json['id']);
    final title = _stringOrNull(json['title']);
    if (id == null || title == null) {
      throw const FormatException('relay returned a task without id or title');
    }
    return Task(
      id: id,
      title: title,
      status: TaskStatus.fromWire(_stringOrNull(json['status'])),
      priority: json['priority'] is int ? json['priority'] as int : 0,
      createdAt: _dateFromSeconds(json['created_at']) ?? DateTime.now().toUtc(),
      updatedAt: _dateFromSeconds(json['updated_at']) ?? DateTime.now().toUtc(),
      channelId: _stringOrNull(json['channel_id']),
      createdBy: _stringOrNull(json['created_by']),
      assignee: _stringOrNull(json['assignee']),
      parentTaskId: _stringOrNull(json['parent_task_id']),
      body: _stringOrNull(json['body']),
      source: _stringOrNull(json['source']),
      sourceRef: _stringOrNull(json['source_ref']),
      dueAt: _dateFromSeconds(json['due_at']),
      doneAt: _dateFromSeconds(json['done_at']),
      archivedAt: _dateFromSeconds(json['archived_at']),
    );
  }

  /// Task id, unique within its community.
  final String id;

  /// Single-line summary of the work.
  final String title;

  /// Lifecycle state.
  final TaskStatus status;

  /// Higher sorts first in the relay's list order.
  final int priority;

  /// When the task was opened.
  final DateTime createdAt;

  /// When the task last changed.
  final DateTime updatedAt;

  /// Channel this task is scoped to, or null for a community-wide task.
  final String? channelId;

  /// Hex pubkey of the author.
  final String? createdBy;

  /// Hex pubkey of the assignee.
  final String? assignee;

  /// Parent task id for a subtask.
  final String? parentTaskId;

  /// Optional Markdown detail.
  final String? body;

  /// Harness or client that opened the task.
  final String? source;

  /// Originating external reference — for a task opened from a thread, the
  /// event id of the message it came from.
  final String? sourceRef;

  /// When the work is due.
  final DateTime? dueAt;

  /// When the task was completed.
  final DateTime? doneAt;

  /// When the task was archived.
  final DateTime? archivedAt;
}

/// One entry in a task's history.
@immutable
class TaskEvent {
  /// Creates a task-event record.
  const TaskEvent({
    required this.id,
    required this.taskId,
    required this.action,
    required this.createdAt,
    this.actor,
    this.fromStatus,
    this.toStatus,
    this.body,
  });

  /// Parses one relay task-event object.
  factory TaskEvent.fromJson(Map<String, dynamic> json) {
    final taskId = _stringOrNull(json['task_id']);
    final action = _stringOrNull(json['action']);
    if (taskId == null || action == null) {
      throw const FormatException('relay returned a malformed task event');
    }
    return TaskEvent(
      id: json['id'] is int ? json['id'] as int : 0,
      taskId: taskId,
      action: action,
      createdAt: _dateFromSeconds(json['created_at']) ?? DateTime.now().toUtc(),
      actor: _stringOrNull(json['actor']),
      fromStatus: _stringOrNull(json['from_status']),
      toStatus: _stringOrNull(json['to_status']),
      body: _stringOrNull(json['body']),
    );
  }

  /// Monotonic event id.
  final int id;

  /// The task this entry belongs to.
  final String taskId;

  /// Raw action string.
  ///
  /// Kept as text rather than an enum because `task_events.action` is
  /// deliberately unconstrained `TEXT` — a new harness must be able to write a
  /// new action without a schema migration, and this client must not choke on
  /// one it has never seen.
  final String action;

  /// When the entry was written.
  final DateTime createdAt;

  /// Hex pubkey of whoever caused the entry.
  final String? actor;

  /// Status before a transition.
  final String? fromStatus;

  /// Status after a transition.
  final String? toStatus;

  /// Free text — a comment, or a persisted summary.
  final String? body;

  /// Whether this entry is the task's persisted summary.
  bool get isSummary => action == TaskEventAction.summaryPersisted.wireValue;
}

/// A task plus its full event history, as returned by `GET /api/tasks/{id}`.
@immutable
class TaskDetail {
  /// Pairs a task with its history.
  const TaskDetail({required this.task, required this.events});

  /// The task.
  final Task task;

  /// Its history, oldest first.
  final List<TaskEvent> events;

  /// The persisted summary entry, if one exists.
  ///
  /// At most one can exist per task — the relay enforces it with a partial
  /// unique index — so callers can use this to decide between "persist" and
  /// "already summarized".
  TaskEvent? get summary {
    for (final event in events) {
      if (event.isSummary) return event;
    }
    return null;
  }
}

/// Builds the JSON body for `POST /api/tasks`.
///
/// Null and blank optional fields are omitted rather than sent as `null`, so
/// the relay's own defaults apply.
///
/// Throws [ArgumentError] when [title] is blank or longer than
/// [maxTaskTitleChars]: the same rejection `validate_title` would return as a
/// 400, raised locally so the sheet can show it without a round trip.
Map<String, Object?> buildCreateTaskPayload({
  required String title,
  String? body,
  String? channelId,
  String? sourceRef,
  String? assignee,
  int? priority,
  DateTime? dueAt,
  String source = 'mobile',
}) {
  final trimmedTitle = title.trim();
  if (trimmedTitle.isEmpty) {
    throw ArgumentError.value(title, 'title', 'must not be empty');
  }
  if (taskTitleLength(trimmedTitle) > maxTaskTitleChars) {
    throw ArgumentError.value(
      title,
      'title',
      'must be at most $maxTaskTitleChars characters',
    );
  }
  final trimmedBody = body?.trim();
  return <String, Object?>{
    'title': trimmedTitle,
    if (trimmedBody != null && trimmedBody.isNotEmpty) 'body': trimmedBody,
    'channel_id': ?channelId,
    'source_ref': ?sourceRef,
    'assignee': ?assignee,
    'priority': ?priority,
    // RFC 3339 in UTC: the relay parses this into `DateTime<Utc>` and only
    // echoes Unix seconds back, so the outbound shape is not the inbound one.
    if (dueAt != null) 'due_at': dueAt.toUtc().toIso8601String(),
    'source': source,
  };
}

/// Resolves the `@handle` labels for a channel's agents.
///
/// Precedence matches the mention pipeline: the agent's own profile display
/// name, then the relay agent directory's, then the first 8 hex characters of
/// its pubkey. All three lookups are lowercase-keyed.
///
/// The result is sorted case-insensitively because [agentPubkeys] arrives as a
/// `Set` — iteration order there is not part of its contract, and an unsorted
/// result would make the composed body unstable between rebuilds.
List<String> resolveAgentHandles({
  required Iterable<String> agentPubkeys,
  required Map<String, String> profileNames,
  required Map<String, String> directoryNames,
}) {
  final handles = <String>[];
  for (final rawPubkey in agentPubkeys) {
    final pubkey = rawPubkey.toLowerCase();
    final profileName = profileNames[pubkey]?.trim();
    final directoryName = directoryNames[pubkey]?.trim();
    final handle = switch ((profileName, directoryName)) {
      (final String name, _) when name.isNotEmpty => name,
      (_, final String name) when name.isNotEmpty => name,
      _ => pubkey.length >= 8 ? pubkey.substring(0, 8) : pubkey,
    };
    if (handle.isNotEmpty && !handles.contains(handle)) handles.add(handle);
  }
  handles.sort((a, b) => a.toLowerCase().compareTo(b.toLowerCase()));
  return handles;
}

/// Combines an optional [body] with an `@`-mention line for [agentHandles].
///
/// Returns null when there is nothing to send, so the caller can omit `body`
/// from the payload entirely rather than posting an empty string.
///
/// The mention line leads because it is the addressing, not the detail: an
/// agent reading the task should see who it is for on the first line.
String? composeTaskBody({String? body, List<String> agentHandles = const []}) {
  final trimmedBody = body?.trim() ?? '';
  final mentions = [
    for (final handle in agentHandles)
      if (handle.trim().isNotEmpty) '@${handle.trim()}',
  ];
  if (mentions.isEmpty) return trimmedBody.isEmpty ? null : trimmedBody;
  final mentionLine = mentions.join(' ');
  return trimmedBody.isEmpty ? mentionLine : '$mentionLine\n\n$trimmedBody';
}

/// Builds the JSON body for `POST /api/tasks/{id}/events`.
///
/// Throws [ArgumentError] for a blank body, which the relay rejects with a 400.
Map<String, Object?> buildTaskEventPayload({
  required TaskEventAction action,
  required String body,
}) {
  final trimmed = body.trim();
  if (trimmed.isEmpty) {
    throw ArgumentError.value(body, 'body', 'must not be empty');
  }
  return <String, Object?>{'action': action.wireValue, 'body': trimmed};
}
